cloudmarker/clouds/gcpcloud.py

```diff
@@ -150,61 +154,23 @@ def read(self):
         Yields:
             dict: Firewall rule or VM instance configuration data.
 
-        Here is an example of a firewall rule record :obj:`dict` yielded
-        by this method:
-
-        .. code:: python
-
-            {
-              "kind": "compute#firewall",
-              "id": "7890789078907890",
-              "creationTimestamp": "2018-12-19T01:43:51.988-08:00",
-              "name": "default-allow-ss",
-              "description": "str",
-              "network": "https://www.googleapis.com/compute/v1/"
-                         "projects/foo/global/networks/default",
-              "priority": 1000,
-              "sourceRanges": [
-                "0.0.0.0/0"
-              ],
-              "targetTags": [
-                "https-server"
-              ],
-              "allowed": [
-                {
-                  "IPProtocol": "tcp",
-                  "ports": [
-                    "80",
-                    "8080-8090"
-                  ]
-                }
-              ],
-              "direction": "INGRESS",
-              "logConfig": {
-                "enable": bool
-              },
-              "disabled": bool,
-              "selfLink": "https://www.googleapis.com/compute/v1/"
-                          "projects/foo/global/firewalls/default-allow-https",
-              "record_type": "firewall_rule"
-            }
-
         """
-        firewall_rules = self._get_firewall_rules()
+        firewalls = self._get_firewalls()
         instances = self._get_instances()
 
-        for rule in firewall_rules:
+        for firewall in firewalls:
             record = {
-                'raw': rule,
+                'raw': firewall,
                 'ext': {
-                    'record_type': 'firewall_rule'
+                    'record_type': 'firewall'
                 },
                 'com': {
                     'cloud_type': 'gcp',
-                    'record_type': 'firewall_rule'
+                    'record_type': None,
```

Comment:
Why you have changed the record_type to `null` here?

Reply:
The properties in the `com` bucket needs to be well thought out. The `com` bucket is something like what CIM of Splunk achieves. It defines standardized property names that would work for all clouds.

In Azure, we have an `nsg` record that has a `security_rules` list with multiple security rules. We take an `nsg` record and split each security rule into its own `firewall_rule` record.

In GCP, we have a `firewall` record (represented as `"kind": "compute#firewall"` in the JSON). This `firewall` record has an `allowed` or `denied` list with multiple port rules. We take a `compute#firewall` record and split each port rule into its own `firewall_rule` record.

In case of `firewall_rule` record, we have complete control on each property in it because we are generating it ourselves. So we are populating its `com` bucket with well thought out names.

But what do we do for the original `nsg` and `compute#firewall` records? They are incompatible with each other. What common `record_type` can we enter that holds good for both? Particularly,

- Should we say `record['com']['record_type']` is `nsg` for Azure but `firewall` for GCP? In that case, someone searching for `record['com']['record_type'] == 'nsg'` would get the Azure NSGs but not the GCP firewall records. Is this good behavior? I don't know.
- Or should we we say `record['com']['record_type']` is `firewall` for both Azure and GCP? In that case a search for `record['com']['record_type'] == 'firewall'` would return both Azure NSGs and GCP firewalls but then there is no other meaningful information about the firewall in the `com` bucket, so what purpose does setting `record['com']['record_type']` to `firewall` for both serve?

In a nutshell, I could not arrive at any good decision for `record['com']['record_type']` value for Azure NSG and GCP firewall. So rather than putting something that may not make sense in one situation or another, I have left it as `None` for now. We can always set it to something useful later provided we can come up with a rationale for it that is consistent for both Azure and GCP.

On the other hand, the `ext` bucket contains record types that makes sense for both Azure and GCP. For example, `record['ext']['record_type']` is `security_rule` for Azure and `firewall` for GCP. The `ext` bucket (extended bucket) is meant for keeping cloud-specific property names and values like this.

The `com` bucket on the other hand is meant for keeping standardized property names and values that work uniformly for all clouds.